co-log/src/Colog/Concurrent.hs

```diff
@@ -0,0 +1,129 @@
+       , convertToLogAction
+       ) where
+
+import Control.Concurrent
+import Control.Concurrent.STM.TBQueue
+
+import Colog.Core.Action (LogAction (..))
+import Colog.Concurrent.Internal
+
+-- $worker-thread
+-- Worker thread approach allows to create a single worker
+-- thread that will perform all logging. This approach allows
+-- user to use different loggers and change them on the flight.
+--
+-- With bracket resource control style code may look as
+-- following:
+--
+-- @
+-- bracket ('mkBackgroundThread' 4096)
+--         ('killBackgroundThread')
+--         $ \thread -> 
+--            let la1 = runInBackgroundThread worker logger1
+--                la2 = runInBackgroundThread worker logger1
+--            unLogger la1 msg1
+--            unLogger la2 msg2
+--            ..
+-- @
+--
+-- This code will run a background thread and process different
+-- loggers in that thread.
+--
+-- You may extend the logger and entire logger will be run
+-- on the background thread. It is not a case with
+-- 'forkBackgroundLogger'.
+
+-- | Create a background worker with a given capacity.
+-- If capacity is reached, then the thread that tries to
+-- write logs will be blocked.
+mkBackgroundThread :: Int -> IO (BackgroundWorker (IO ()))
+mkBackgroundThread capacity = do
+  queue <- newTBQueueIO (fromIntegral capacity)
+  tid <- forkIO $ forever $ join $ atomically $ readTBQueue queue
+  pure $ BackgroundWorker tid (writeTBQueue queue)
+
+-- | Run logger action asynchronously in the worker thread.
+-- Logger is executed in the other thread entirely, so if
+-- logger takes any thread related context it will be
+-- read from the other thread.
+runInBackgroundThread :: BackgroundWorker (IO ()) -> LogAction IO msg -> LogAction IO msg
+runInBackgroundThread bt logAction = LogAction $ \msg -> do
+  atomically $ backgroundWorkerWrite bt $ unLogAction logAction msg
+
+-- $background-logger
+--
+-- Background logger is specialized version of the Worker 
+-- process. Instead of running any user's logger it will
+-- accept @msg@ type instead and process it with a single
+-- logger defined at creation time. This approach may be more
+-- efficient comparing to the 'runInBackgroundThread' but it's less
+-- generic.
+--
+-- @
+-- bracket ('forkBackgroundLogger' 4096 logger1)
+--         ('killBackgroundThread')
+--         $ \logger-> 
+--            let la = convertToLogAction logger
+--            unLogger la msg1
+--            ..
+-- @
+--
+-- You may extend the logger and but all extensions will be run
+-- in the user thread, only @logger1@ is run in background.
+-- It is not a case with 'runInBackgroundThread'.
+
+-- | Creates background logger with given @Capacity@,
+-- takes a 'LogAction' that should describe how to write
+-- logs.
+--
+-- @capability@ - parameter tells how many in flight messages
+-- are allowed, if that value is reached then user's thread
+-- that emits logs will be blocked until any message will be
+-- written.
```

Comment:
I think alternative strategy is to drop all log messages after reaching the limit. But I agree with making current behavior the default one. Probably in production it's more important to lose some performance and hang for a while rather than losing important logs about what is actually going on within the running system.

Reply:
I think I'll add some docs and expose more internal so users will be able to define such strategies on their own.  I don't like dropping messages, it should be some very special use-case. If you have too many logs so you can't dump them, it means that your environment has some severe problems that should be solved anyway.